lib/auth/oidc/client_assertion_test.go

```diff
@@ -0,0 +1,520 @@
+				OIDCClientAssertion: &structs.OIDCClientAssertion{
+					KeySource:    structs.OIDCKeySourceClientSecret,
+					KeyAlgorithm: "HS256",
+					Audience:     []string{"test-audience"},
+					ExtraHeaders: map[string]string{
+						"test-header": "test-value",
+					},
+				},
+			},
+			wantErr:     true,
+			expectedErr: `invalid secret length for algorithm: "HS256" must be at least 32 bytes long`,
+		},
+		{
+			name: "invalid client secret kid in extra header",
+			config: &structs.ACLAuthMethodConfig{
+				OIDCClientID:     "test-client-id",
+				OIDCClientSecret: "1234567890abcdefghijklmnopqrstuvwxyz",
+				OIDCClientAssertion: &structs.OIDCClientAssertion{
+					KeySource:    structs.OIDCKeySourceClientSecret,
+					KeyAlgorithm: "HS256",
+					Audience:     []string{"test-audience"},
+					ExtraHeaders: map[string]string{
+						"kid": "test-kid",
+					},
+				},
+			},
+			wantErr:     true,
+			expectedErr: `WithHeaders: "kid" not allowed in WithHeaders; use WithKeyID instead`,
+		},
+		{
+			name: "invalid key algorithm none",
+			config: &structs.ACLAuthMethodConfig{
+				OIDCClientID:     "test-client-id",
+				OIDCClientSecret: "1234567890abcdefghijklmnopqrstuvwxyz",
+				OIDCClientAssertion: &structs.OIDCClientAssertion{
+					KeySource:    structs.OIDCKeySourceClientSecret,
+					KeyAlgorithm: "none",
+					Audience:     []string{"test-audience"},
+					ExtraHeaders: map[string]string{
+						"test-header": "test-value",
+					},
+				},
+			},
+			wantErr:     true,
+			expectedErr: `unsupported algorithm "none" for client secret`,
+		},
+		{
+			name: "invalid key algorithm None",
+			config: &structs.ACLAuthMethodConfig{
+				OIDCClientID:     "test-client-id",
+				OIDCClientSecret: "1234567890abcdefghijklmnopqrstuvwxyz",
+				OIDCClientAssertion: &structs.OIDCClientAssertion{
+					KeySource:    structs.OIDCKeySourceClientSecret,
+					KeyAlgorithm: "None",
+					Audience:     []string{"test-audience"},
+					ExtraHeaders: map[string]string{
+						"test-header": "test-value",
+					},
+				},
+			},
+			wantErr:     true,
+			expectedErr: `unsupported algorithm "None" for client secret`,
+		},
+		// expected non-nil error; got nil
+		{
+			name: "invalid missing audience",
+			config: &structs.ACLAuthMethodConfig{
+				OIDCClientID:     "test-client-id",
+				OIDCClientSecret: "1234567890abcdefghijklmnopqrstuvwxyz",
+				OIDCClientAssertion: &structs.OIDCClientAssertion{
+					KeySource:    structs.OIDCKeySourceClientSecret,
+					KeyAlgorithm: "HS256",
+					ExtraHeaders: map[string]string{
+						"test-header": "test-value",
+					},
+				},
+			},
+			wantErr:     true,
+			expectedErr: "missing Audience",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.config.Canonicalize() // inherits ClientSecret from OIDCClientAssertion
+			jwt, err := BuildClientAssertionJWT(tt.config, nil, "")
+			if tt.wantErr {
+				must.Error(t, err)
+				must.StrContains(t, err.Error(), tt.expectedErr)
```

Comment:
```suggestion
				must.ErrContains(t, err, tt.expectedErr)
```

And similar for all the tests in this file.

You could also swap "wantErr" out for an empty string check on `tc.expectedErr`, but up to you.